[Roach] 미션 4: 모든 기물 배치하기

tmdgusya opened the pull request:
미션 4 는 조금의 학습 용도로 작성된것도 있습니다.
PieceList 같은 경우도 Board 에서 List 에서 말을 찾거나 하는 과정들이 포함되어 있기에, 이런 책임도 보드가 하는것은 아닌것 같다고 생각하여 
PieceList 로 분할하였습니다. 그래서 Board 에서는 PieceList 의 메소드들을 이용해서 간단하게 원하는 말들을 찾을 수 있도록 하였습니다.

StringUtils 의 append 같은 경우도 static method 로 이용할 수 있도록 요구사항에 적혀있어 필드또한 static 하게 하였고,
이를 print() 할때 `getBoardStatusToString` 이 작동되어서 중복이 되는 경우가 있어 init() 으로 초기화 할 수 있도록 하였습니다.

뭔가 나눌때 분리하다보니, 아직 미숙한 점을 많이 느끼는 것 같습니다.
1. init() 과 같은 초기화 메소드 없이는 불가능한가? => 이점은 static 이 아니면 가능할 것 같기도 한데 아직은 잘 모르겠습니다.
2. PieceList 를 나눴을때 이점은 ? Board 안에서 List 에서 특정 조건으로 찾는 로직들이 Board 보다는 PieceList 에서 관리하는게 맞다고 생각했고, 조건탐색 로직등의 코드를 덜어낼 수 있었음 => 근데 Board 에 있었을때가 나은가? PieceList 로 나눠졌을때가 더 나은가? 라고 했을때는 편의성 ? 유지보수 ? 이정도 말고는 이점을 생각해 내긴 힘들었습니다. 다만 저 이점 두개로도 나눌만 하다고는 생각을 하긴 하였는데.. 아직 옳은 것인지는 잘 모르겠습니다.

tmdgusya committed:
refactor : \n 의 반복 코드 제거

StringUtils.java

- NEWLINE : 변수를 line.seperator 를 시스템에서 받아오도록 함.
- appendNewLine : 내부에 StringBuilder 객체를 두어서 해당 객체를 통해 뛰어쓰기 연산을 하도록 해줌.
- convertString : 현재까지 더해온 StringBuilder 를 String 형태로 리턴시켜줌.

Board.java

- StringUtils 사용하여 중복 코드 제거

tmdgusya committed:
feat : Piece 로 Rename 및 팩토리 메소드를 통한 Piece 생성

Piece.java

- static 변수로 각 Piece 마다 representation 을 설정해 주었음.
- private 생성자로 두어 Piece 는 수정이 불가능하게 하여 값 객체로서 생성함.

PieceTest.java

- 각 말별로 테스트를 생성하였음 -> 후에 분리될 수도 있다고 생각하여 각 말별로 일단 테스트를 나눠하는 방식으로 진행하였음.

tmdgusya committed:
feat : 전체 기물의 상태를 볼 수 있는 체스판 구현

StringUtils.java

 - append 시 StringBuilder 를 return 하게 해서 add 할때 연달아 붙어서 할 수 있도록 하였음(편의성을 위해서)

 - init() : static 으로 하다보니 객체 자체를 초기화 할일이 생기는데 그때 사용하려고 만들어 두었음.

 PieceList.java

 - 내부에서 arrayList 를 이용하였고, 각 메소드가 Board 에서 많아지는 걸 방지하기 위해 랩퍼 클래스로 분할하였음

 - add : 이것또한 편의적으로 이어 붙이기 위해서 객체를 리턴하도록 함.

 - getPieceList() : 외부에서 객체를 거치지 않고 수정 / 삽입하는 것을 방지하기 위해 unmodifiableList 로 넘겨주었음.

 - getStringPiecesList() : 시작인덱스부터, 끝 인덱스까지 찾는것인데, 주요 말을 찾을때 사용하려고 만들어둔 편의 메소드

 Board.java

 - List 부분 PiceList 사용

 - getPiecesExceptPawns() : 각 List 의 getStringPiecesList 메소드를 통하여 startIndex, endIndex 로 해당 부분 구하도록 함.

 - addPiecesToBoardSize() : 새롭게 생긴말들을 초반에 추가해주는 작업.

tmdgusya committed:
test : BoardTest.java

- Board 부분 코드 업데이트로 인한 수정

tmdgusya committed:
feat : 검은색 말과 흰색 말을 구분할 수 있는 메소드 추가

- 현재 color 상태값과 Enum 을 비교하여 boolean 으로 리턴하도록 했음.

tmdgusya committed:
feat : 검은색 말과 흰색 말을 구분할 수 있는 메소드 테스트

- 현재 color 상태값과 Enum 을 비교하여 boolean 값이 정상적으로 색깔을 구분하는지 테스트

tmdgusya committed:
refactor : 중복코드 및 코드 정렬

Board.java

- 코드 정렬 및 static import 정확하게 명시

PieceList

- 코드 정렬 및 불필요한 import 삭제

tmdgusya commented on src/main/java/net/honux/chess/util/PieceList.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public List<Piece> getPieceList() {
+        return Collections.unmodifiableList(pieceList);
```
unmodifiableList 로 list 를 넘겨주는 이유는 외부에서 add remove 등의 메소드로 접근하는 일이 없도록 하기 위해서 아래와 같이 선언했습니다. 객체를 통해서만 list 에 접근할 수 있도록 했습니다.

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,120 @@
 
-    private static final int BOARD_SIZE = 8;
+    public static final int BOARD_SIZE = 8;
     private static final String BLANK = "********";
     private static final int initialSpaceInterval = 4;
```
이런 정보들은 지역변수로 만들어서 사용하는건 어떨까요?

tmdgusya replied:
그게 나을 것 같네요. 굳이 static 변수로서 있을 이유가 없는 것 같습니다.
단순히 메소드 시에만 생성되고 사라지는 지역변수로서 있으면 메모리상으로도 훨씬 좋을 것 같네요.

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,120 @@
+    private PieceList whitePieceList;
+    private PieceList blackPieceList;
 
     public void initialize() {
-        whitePawnList = new ArrayList<>();
-        blackPawnList = new ArrayList<>();
-        addPawnToBoardSize();
+        whitePieceList = new PieceList();
+        blackPieceList = new PieceList();
+        addPiecesToBoardSize();
     }
```
`PieceList`라는 클래스를 만드셨네요.
우선 클래스 이름에 `List`가 붙는거 보다 `Pieces`가 어떨까 하는 생각이 들구요.

제가 보기에는 지금 Board 클래스가 너무 좀 지저분하다? 장황하다?
`Board` 클래스를 같이 수정해 봅시다.

여기 코드를 보면 흰색 피스, 검은색 피스로 나눠서 관리를 하시겠다는 의도인거 같은데
단순히 흰색, 검은색으로만 나누기 보다는 
흰색, 검은색, 흰색 폰, 검은색 폰으로 나누는게 어떨까 하는 생각이 드네요.
그러면 `Piece` 클래스에 `getStringPiecesList(int startIndex, int endIndex)` 같은 메서드가 필요가 없어지겠죠?
밑에서도 피드백으로 드리겠지만 이부분이 굉장히 지저분해 보였거든요.

그리고 이왕에 `PieceList`를 만드셨다면 만드신 클래스를 적극 활용하는게 좋을거 같습니다.
이와 관련된 피드백은 밑에다 남길게요.

tmdgusya replied:
색깔 별로 분리하려했는데 Pawn 도 따로 분리하는 것이 깔끔할것 같군요 감사합니다!

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,120 @@
     }
 
     public String getBlackPawnsResult() {
-        StringBuilder sb = new StringBuilder();
-        for (Pawn pawn : blackPawnList) {
-            sb.append(pawn.getRepresentation());
-        }
-        return sb.toString();
+        return blackPieceList.getStringPiecesList(8, 15);
+    }
+
+    public String getBlackPiecesExceptPawns() {
+        return blackPieceList.getStringPiecesList(0, 7);
+    }
+
+    public String getWhitePiecesExceptPawns() {
+        return whitePieceList.getStringPiecesList(8, 15);
     }
```
이렇게 숫자를 넘겨서 String을 받아 오는거 보다는 위에서 언급한거 처럼
차라리 `PieceList`를 4개 만들고 
```java
public String getWhitePawnsResult() {
    return whitePieceList.getString()
}
```
뭐 이런식으로 반환하는 메서드 4개를 만드는게 더 깔끔할거 같은데 어떻게 생각하시나요~

tmdgusya replied:
그것도 좋은 방법일거 같고 , 하나 떠오르는건 아래와 같이 해서 list 를 받아서 해도될것 같네요!
```java
private String getStringPiecesResult(Pieces pieces) {
        StringBuilder sb = new StringBuilder();
        for(Piece p : pieces.getPieceList()){
            sb.append(p.getRepresentation());
        }
        return sb.toString();
    }
```

tmdgusya replied:
아 상코 말대로 Pieces 에 getString() 으로 해서 하니 더깔끔해지네요.
```java
public String getString() {
        StringBuilder sb = new StringBuilder();
        for(Piece p : this.pieceList){
            sb.append(p.getRepresentation());
        }
        return sb.toString();
    }

private String getStringPiecesResult(Pieces pieces) {
        return pieces.getString();
    }
```

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,120 @@
+        appendNewLine(getBlackPiecesExceptPawns());
+        appendNewLine(getBlackPawnsResult());
         for (int i = 0; i < initialSpaceInterval; i++) {
-            sb.append(BLANK).append("\n");
+            appendNewLine(BLANK);
         }
-        sb.append(getWhitePawnsResult()).append("\n");
-        sb.append(BLANK);
-        return sb.toString();
+        appendNewLine(getWhitePawnsResult());
+        appendNewLine(getWhitePiecesExceptPawns());
+        return convertString();
     }
 
     public void print() {
+        StringUtils.init();
         System.out.println(this.getBoardStatusToString());
     }
```
> print() 할때 getBoardStatusToString 이 작동되어서 중복이 되는 경우가 있어 init() 으로 초기화 할 수 있도록 하였습니다.

`StringUtils` 클래스에 `StringBuilder`가 있군요.
유틸성 클래스에서 클래스 변수를 가지게 하지말고
`getBoardStatusToString()`에서 `StringBuilder`를 지역변수로 가진다면
`print()` 메서드도 없앨 수 있을거 같은데요?
간단히 문자열을 합치는 거지만 `StringBuilder`를 사용하고 싶으시면
파라미터로 `StringBuilder`를 넘길 수도 있겠죠?

tmdgusya replied:
오 좋은 방법인것 같네요. 정적 메소드를 이용할때 지역변수의 이용방법을 숙지해둬야겠네요.

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,120 @@
-    private void addPawnToBoardSize() {
+    private void addPiecesToBoardSize() {
+        setBlackPiecesExceptPawns();
         for (int i = 0; i < BOARD_SIZE; i++) {
-            whitePawnList.add(new Pawn(Color.WHITE));
-            blackPawnList.add(new Pawn(Color.BLACK));
+            blackPieceList.add(createBlackPawn());
+            whitePieceList.add(createWhitePawn());
         }
+        setWhitePiecesExceptPawns();
+    }
+
+    private void setWhitePiecesExceptPawns() {
+        whitePieceList.add(createWhiteRook()).add(createWhiteKnight())
+                .add(createWhiteBishop()).add(createWhiteQueen()).add(createWhiteKing())
+                .add(createWhiteBishop()).add(createWhiteKnight()).add(createWhiteRook());
+    }
+
+    private void setBlackPiecesExceptPawns() {
+        blackPieceList.add(createBlackRook()).add(createBlackKnight())
+                .add(createBlackBishop()).add(createBlackQueen()).add(createBlackKing())
+                .add(createBlackBishop()).add(createBlackKnight()).add(createBlackRook());
     }
```
기왕에 만든거 'PieceList'를 적극 활용해 봅시다!
`PieceList` 그냥 `Pieces` 라고 할게요. 

`Pieces`에  정적 팩토리 메서드들을 만들어 보는건 어떨까요?
그래서 `Board` 클래스에서 `Pieces` 클래스에게 요청을 하는거죠.
'리스트에 검은색 폰 8개 넣어서 반환해줘' 뭐 이런식으로

제 설명이 좀 이해가 안되실 수도 있을거 같은데 댓글로 우리 소통합시다.
이해 안되시는거 있으시면 댓글 남겨주세요~

정적 팩토리 메서드는 이펙티브 자바 [아이템 1. 생성자 대신 정적 팩터리 메서드를 고려하라.] 참고 하셔도 좋을거 같네요.
이것도 도움이 되실거 같아서 링크 남깁니다. [일급 컬렉션 (First Class Collection)의 소개와 써야할 이유](https://jojoldu.tistory.com/412)

tmdgusya replied:
아래와 같이 진행하였습니다. 저 링크글은 진짜 좋은글이네요! :)

이렇게 하니 Pices 를 만든 진짜 목적 그대로 이용되고 있는 것 같네요. 메인 로직에서의 분리와 List 단위의 반복작업이 분리된것 같네요!
해당 부분은 조금 더 공부해봐야할듯 합니다!

```java
private void addPiecesToBoardSize() {
        blackPieces = Pieces.createBlackPiecesExceptPawns();
        blackPawns = Pieces.createBlackPawns();
        whitePawns = Pieces.createWhitePawns();
        whitePieces = Pieces.createWhitePiecesExceptPawns();
    }
```

SANGCO commented on src/main/java/net/honux/chess/util/PieceList.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public List<Piece> getPieceList() {
+        return Collections.unmodifiableList(pieceList);
```
> 객체를 통해서만 list 에 접근할 수 있도록 했습니다.

객체를 통해서만 list를 수정할 수 있게 의도 하셨다는 거죠?

tmdgusya replied:
네 객체를 통해서만 접근이 가능해야 나중에 검증로직같은 것을 거칠수 있을것이고,
Collection을 클래스로 랩핑한 이유가 합당하다고 생각되었습니다.

tmdgusya replied:
비즈니스 로직을 Pieces 로 옮기다보니 getPieceList() 라는 메소드도 없어지게 되었습니다. 
기존에는 iterator 를 사용하기 위해 사용했는데, 정적 팩토리 메소드로 Pieces 에서 Pieces 를 만들어주게되면서 필요없어지게 되어서 삭제하였습니다

SANGCO commented on src/main/java/net/honux/chess/util/StringUtils.java:
```diff
@@ -0,0 +1,26 @@
+public class StringUtils {
+
+    public static final String NEWLINE = System.getProperty("line.separator");
+    private static StringBuilder sb = new StringBuilder();
```
위에서 언급했던 부분입니다.
유틸 클래스에 있을 필요가 있을까요?

tmdgusya replied:
StringBuilder 는 필요없을 것 같으나, NEWLINE 은 Util 클래스에 있어도 된다고 생각합니다.

SANGCO commented on src/main/java/net/honux/chess/util/StringUtils.java:
```diff
@@ -0,0 +1,26 @@
+        return sb.toString();
+    }
+
+    public static void init() {
+        sb = new StringBuilder();
+    }
```
이 메서드도 리팩토링 과정에서 쓸모가 없어지지 않을까 싶은데요.

tmdgusya replied:
네 리뷰과정을 반영하다보니 사라지게 되었네요.

SANGCO commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -22,52 +21,55 @@ void generateBoard() {
-                () -> assertThat(board.findWhitePawn(8).get()).isEqualTo(white),
-                () -> assertThat(board.blackPawnSize()).isEqualTo(9),
-                () -> assertThat(board.findBlackPawn(8).get()).isEqualTo(black)
+                () -> assertThat(board.whitePawnSize()).isEqualTo(Board.BOARD_SIZE*2+1),
+                () -> assertThat(board.findWhitePawn(Board.BOARD_SIZE*2).get()).isEqualTo(whitePawn),
+                () -> assertThat(board.blackPawnSize()).isEqualTo(Board.BOARD_SIZE*2+1),
+                () -> assertThat(board.findBlackPawn(Board.BOARD_SIZE*2).get()).isEqualTo(blackPawn)
```
`Board.BOARD_SIZE * 2 + 1`
중간에 공백을 다 두셔야 합니다.

tmdgusya replied:
네 가독성을 위해서 떨어뜨리는게 좋겠군요.
세심한 부분 리뷰 감사합니다 👍

SANGCO commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -22,52 +21,55 @@ void generateBoard() {
     void checkPawnRepresentationOnTheBoard() {
         board.initialize();
         Assertions.assertAll(
+                () -> assertThat(board.getBlackPiecesExceptPawns()).isEqualTo("RNBQKBNR"),
+                ()->assertThat(board.getWhitePiecesExceptPawns()).isEqualTo("rnbqkbnr"),
                 () -> assertThat(board.getWhitePawnsResult()).isEqualTo("pppppppp"),
                 () -> assertThat(board.getBlackPawnsResult()).isEqualTo("PPPPPPPP")
         );
```
"pppppppp", "PPPPPPPP", "RNBQKBNR", "rnbqkbnr"
요런거 테스트 클래스 내부에서 상수로 빼서 사용하면 더 좋지 않을까요?

tmdgusya replied:
## 리뷰 과정 중 checkBoardPrintMessage() 만 유지

- 원래 각 Pieces 의 Result 도 보는 Test 가 이루어졌으나, checkBoardPrintMessage() 를 통해서 한번에 테스트가 가능하다는 것을 알게되었습니다. 그래서 해당 부분 테스트를 하나로 통일하였고, 코드가 변경되면서 아직까지는 상수로 빼진 않아도 될듯합니다.

SANGCO reviewed (changes requested):
Roachs님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

tmdgusya committed:
refactor : 코드리뷰 내용 반영

Board.java
- Pawn 과 다른 Pieces 를 구분하는 로직을 없애고 따로 List 를 두어 관리하는게 편리하다고 생각하여 4개의 리스트를 생성하였음.
- Pieces 를 만든만큼 List 로직을 Pieces 에서 받아와서 getStringPiecesResult() 를 이용할 수 잇도록 하였음.

Pieces.java
- PieceList -> Pieces 로 refactor name( List 보다는 복수형으로 기술하는것이 낫다고 생각됨)
- createPawns() : Pieces 를 분리한 만큼 정적 팩토리 메소드로 해당하는 Pieces() 를 리턴하도록 하였음. 보드에서 하는것 보다는 Pieces 에서 이뤄지는 것이 맞다고 판단됨.
- getString() : 현재 객체의 List 에 있는 Pieces 들의 representation 을 String 으로 리턴해주는 함수.

StringUtils.java
- StringBuilder 를 외부에서 받아오도록 해서 클래스 자체에서 StringBuilder 를 가지지 않도록 했음.

tmdgusya committed:
test : 코드리뷰 내용 반영

BoardTest.java

- 기존에 있던 checkPawnRepresentationOnTheBoard() 는 checkBoardPrintMessage() 에서 같이 테스트 해도 무방하다고 생각하여 해당 테스트를 삭제하고 통일시켰음.

StringUtilsTest.java
- appendNewLine() 테스트를 통하여 NEWLINE 이 추가가 잘되는지 확인하였음.

tmdgusya committed:
refactor : 코드리뷰 내용 반영

- Pieces.java

getPieceList() 는 초반에 Board 에서 iterator 를 돌려서 사용할때 이용했는데, 해당 비즈니스 로직을 Pieces 로 옮기게 되면서 삭제하게 되었습니다.

tmdgusya commented:
@SANGCO 리뷰 반영 내용을 커밋에 담아서 반영하였습니다.

1급 컬렉션 관련 내용은 상당히 흥미있었습니다. 
앞으로 많이 애용해야겠다는 생각이 들었습니다. 유지보수와 중복코드 모두 챙길수 있는 이점이 있다고 생각되었습니다. 
👍

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,98 @@
-        if (pawn.getColor() == Color.BLACK) {
-            return blackPawnList;
+    private Pieces getListByColor(Piece piece) {
+        if (piece.getColor() == Color.BLACK) {
+            return blackPieces;
         }
-        return whitePawnList;
+        return whitePieces;
+    }
```
이 private 메서드를 사용하는 메서드 밑으로 이동시켜 주세요~

SANGCO replied:
whitePawns, blackPawns에 넣을 때는 어떻게 하나요?

tmdgusya replied:
그 부분은 representation 으로 구분지어야 할듯하네요.

tmdgusya replied:
```java
 private Pieces getListByColor(Piece piece) {
        if (piece.isBlack()) {
            if(piece.getRepresentation() == BLACK_PAWN_REPRESENTATION) {
                return blackPawns;
            }
            return blackPieces;
        }
        if (piece.isWhite() && piece.getRepresentation() == WHITE_PAWN_REPRESENTATION) {
                return whitePawns;
        }
        return whitePieces;
    }
```

tmdgusya replied:
```java
        if (piece.isBlack() && piece.getRepresentation() == BLACK_PAWN_REPRESENTATION) {
            return blackPawns;
        }
        if (piece.isWhite() && piece.getRepresentation() == WHITE_PAWN_REPRESENTATION) {
            return whitePawns;
        }
        return piece.isBlack() ? blackPieces : whitePieces;
```

tmdgusya replied:
두가지 중 아래 코드는 간결하나 else 가 사용된다는 단점이 있는것 같은데 SANGCO 는 어떤 방식이 좋다고 생각하시나요?

tmdgusya replied:
저는 개인적으로 위의 코드가 괜찮다고 생각되는데 한번 견해를 들어보고 싶습니다 :)

SANGCO replied:
저는 if문을 제거 하기 위해 메서드 분리를 했을거 같아요.
add()라는 하나의 메서드로 모든 종류의 piece를 추가하려고 하니깐
저런 if문들이 발생한게 아닌가 생각이 드네요.

이번 미션에서는 우선 밑에 거로 해서 진행하시고
향후에 미션을 진행 하시면서 메서드가 점점 복잡해 진다고 느껴지시면
적극적인 메서드 분리를 시도해 봅시다!

SANGCO commented on src/main/java/net/honux/chess/util/Pieces.java:
```diff
@@ -0,0 +1,66 @@
+package net.honux.chess.util;
```
Pieces가 유틸성 클래스 일까요?

tmdgusya replied:
List 라 생각해서 유틸성 클래스라 생각했는데 혹시 어떤 패키지라고 생각되시나요?

SANGCO replied:
core 패캐지에 board 클래스 옆이 더 어울릴거 같네요~

SANGCO commented on src/main/java/net/honux/chess/util/Pieces.java:
```diff
@@ -0,0 +1,66 @@
+        return pieces;
+    }
+
+    public static Pieces createWhitePiecesExceptPawns() {
+        return new Pieces().add(createWhiteRook()).add(createWhiteKnight())
+                .add(createWhiteBishop()).add(createWhiteQueen()).add(createWhiteKing())
+                .add(createWhiteBishop()).add(createWhiteKnight()).add(createWhiteRook());
+    }
```
이거는 피드백은 아니고 제 개인적인 스타일인데 
한줄로 세우는게 더 보기 깔끔하지 않나요?ㅋ

tmdgusya replied:
지금 한번해봤는데 한줄로 세우는 게 깔끔해 보이긴하네요.
딱 접기 누르니까 뭔가 그럴 듯 하네요 :)

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,98 @@
-        whitePawnList = new ArrayList<>();
-        blackPawnList = new ArrayList<>();
-        addPawnToBoardSize();
+        whitePieces = new Pieces();
+        blackPieces = new Pieces();
+        whitePawns = new Pieces();
+        blackPawns = new Pieces();
+        addPiecesToBoardSize();
```
```java
whitePieces = createWhitePiecesExceptPawns();
blackPieces = createBlackPiecesExceptPawns();
whitePawns = createWhitePawns();
blackPawns = createBlackPawns();
```

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,98 @@
-            whitePawnList.add(new Pawn(Color.WHITE));
-            blackPawnList.add(new Pawn(Color.BLACK));
-        }
+    private void addPiecesToBoardSize() {
+        blackPieces = Pieces.createBlackPiecesExceptPawns();
+        blackPawns = Pieces.createBlackPawns();
+        whitePawns = Pieces.createWhitePawns();
+        whitePieces = Pieces.createWhitePiecesExceptPawns();
     }
```
없어도 되겠죠?

tmdgusya replied:
아 위의 답글을 보니 없어도 되겠군요. 괜한 코드 중복.. 지적해주셔서 감사합니다.

SANGCO commented on src/main/java/net/honux/chess/util/Pieces.java:
```diff
@@ -0,0 +1,66 @@
+
+    private final List<Piece> pieceList = new ArrayList<>();
+
+    public Pieces add(Piece piece) {
+        pieceList.add(piece);
+        return this;
+    }
```
이 구조면 add()만 계속 호출하면 한도 끝도 없이 List가 길어 질수도 있겠네요.

tmdgusya replied:
흐음 그렇다면 close 같은 함수를 호출하는게 좋을까요? 아니면 그냥 add 에서 this 를 호출하지 못하도록 할까요?
설계 목적상 계속 호출하려고 한것도 있어서.. 이부분은 어떻게 처리해야할지 감이 안오네요

tmdgusya replied:
```java
 public Pieces add(Piece piece) {
        pieceList.add(piece);
        if(!validateSize()) {
            throw new IndexOutOfBoundsException("더 이상 Piece 를 추가할 수 없습니다.");
        }
        return this;
    }
```

tmdgusya replied:
위와 같이 바꾸어 보았는데 어떨까요?

SANGCO replied:
현재 단계에 미션에서는 하나의 Pieces 클래스가 가지고 있는 List에
8개가 딱 들어가야 한잖아요?
그래서 지금 수정하신거 보면 add() 하고 if문을 체크 하시는데
add() 하기 전에 간단하게 체크하는게 있으면 좋겠다 싶어서 리뷰 남겼답니다.

tmdgusya replied:
감사합니다. 생각해보니 유효성검증이 먼져 진행되어야 하는데 제가 밑에 적은것 같네요.

SANGCO commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -19,55 +19,29 @@ void generateBoard() {
-                "********\n" +
-                "pppppppp\n" +
-                "********";
+                "RNBQKBNR" + StringUtils.NEWLINE +
```
요런거는 StringUtils static import 하면 깔끔하게 쓸 수 있습니다!

SANGCO reviewed (changes requested):
몇 가지 추가로 리뷰 남깁니다~

tmdgusya committed:
refactor : 2차 코드리뷰 반영

Board.java

- addPiecesToBoardSize() : 중복코드로 판단되어 제거하였음.
- getListByColor() : 해당 piece 에 맞는 List 를 리턴하기 위에 조건문 추가, naming 복수형인 getPiecesByColor 로 변경하였음.
- findWhitePiece / findBlackPiece : List 4개 에 따른 메소드 분리 구현

Pieces.java
- add() : validateSize() 를 통해서 8 이상일때 추가시 IndexOutOfBoundsException 를 리턴하도록 하였음.

BoardTest.java / StringUtilsTest.java
- StringUtils static import 하여 가독성을 높임

tmdgusya commented on src/test/java/net/honux/chess/util/StringUtilsTest.java:
```diff
@@ -4,13 +4,14 @@
-        StringUtils.appendNewLine(sb, "test");
-        assertThat(StringUtils.convertString(sb)).isEqualTo("test"+StringUtils.NEWLINE);
+        appendNewLine(sb, "test");
+        assertThat(convertString(sb)).isEqualTo("test"+NEWLINE);
```
다음 머지때 띄어쓰기 해서 올리기

SANGCO commented on src/main/java/net/honux/chess/core/Pieces.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public static Pieces createWhitePiecesExceptPawns() {
+        return new Pieces().add(createWhiteRook()).add(createWhiteKnight()).add(createWhiteBishop()).add(createWhiteQueen()).add(createWhiteKing()).add(createWhiteBishop()).add(createWhiteKnight()).add(createWhiteRook());
```
.add(createWhiteRook())
.add(createWhiteRook())
.add(createWhiteRook())
.add(createWhiteRook())

제가 좋아하는 스타일이란건 요건식으로
표현을 애매하게 했군요ㅠㅠ

SANGCO replied:
원래 하시던데로 하셔도 될거 같아요.
근데 지금 처럼 길게 하는건 아닌거 같아요ㅋ

SANGCO commented on src/main/java/net/honux/chess/core/Pieces.java:
```diff
@@ -0,0 +1,66 @@
+        return sb.toString();
+    }
+
+    private boolean validateSize() {
+        return this.size() < 8;
+    }
```
마이너스 값 까지 고려해서
현재는 
`if (!validateSize()) {`
이런데 
`if (validateSize()) {`
이렇게 될 수 있도록 조건문을 만들어보면 어떨까요?

tmdgusya replied:
네넵 ! 연산자를 쓰지 않고 기본 조건을 쓰는게 좋겠군요
```java
 private boolean validateSize() {
        return this.size() >= 8;
    }
```

SANGCO replied:
8은 들어가면 안되지 않나요?

음 메서드 명을 좀 변경해야 할거 같네요.
지금 메서드 명으로는 원래 하신거 처럼
`if (!validateSize()) {`
이게 더 어울리는거 같네요.

tmdgusya replied:
네넵 유효하지 않은 사이즈일때가 좋은 표현방식인것 같네요 감사합니다~!

SANGCO replied:
```java
    private Pieces add(Piece piece) {
        if (isValid()) {
            pieceList.add(piece); 
        }
        return this;
    }
```

익셉션은 isValid 메서드에서 던지는 걸로 하면 어떨까요?
그리고 현재 add() 메서드가 validateSize()를 사용하고 있는데
validateSize() 메서드가 add() 메서드 위에 있네요.

https://stackoverflow.com/a/32131711

이거 링크 한번 참고해주세요.

tmdgusya replied:
```java
private void validateSize() {
        if(this.size() >= 8) {
            throw new IndexOutOfBoundsException("더 이상 Piece 를 추가할 수 없습니다.");
        }
    }


    private Pieces add(Piece piece) { 
        validateSize();
        pieceList.add(piece);
        return this;
    }
```

이 방식이 더 나아보인다고 생각하는데 어떻게 생각하시나요?

SANGCO commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,102 +1,110 @@
-        for (int i = 0; i < BOARD_SIZE; i++) {
-            whitePawnList.add(new Pawn(Color.WHITE));
-            blackPawnList.add(new Pawn(Color.BLACK));
+    private Pieces getPiecesByColor(Piece piece) {
+        if (piece.isBlack() && piece.getRepresentation() == BLACK_PAWN_REPRESENTATION) {
+            return blackPawns;
+        }
+        if (piece.isWhite() && piece.getRepresentation() == WHITE_PAWN_REPRESENTATION) {
+            return whitePawns;
         }
+        return piece.isBlack() ? blackPieces : whitePieces;
     }
```
add() 메서드가 사용하고 있으니깐 그 밑으로 이동시켜 주세요.

tmdgusya replied:
네넵 연간되는 것이라 근처에 둬야 가독성이 좋겠군요.

SANGCO reviewed (changes requested):
간단한거 몇가지만 더 수정하고 후딱 다음 미션으로 넘어갑시다~

tmdgusya committed:
refactor : 3차 코드리뷰 반영

1. Board.java
- getPiecesByColor() : 가독성을 위한 add() 밑으로 자리 이동

2. Pieces.java
- createWhitePiecesExceptPawns() / createBlackPiecesExceptPawns : 가독성을 위해 한줄로 작성
- validateSize() : ! 를 쓰지 않도록 하기 위해 예외상황을 true 조건으로 설정

tmdgusya committed:
refactor : 3차 코드리뷰 반영

1. StringUtilsTest.java

- 가독성을 위한 띄어쓰기 작성

tmdgusya committed:
fix : 3차 코드리뷰 반영

1. Board.java
- add() : Method 삭제 더이상 Board 단에서는 추가될 일이 없을 것이라고 생각됨. 이로써 1급 콜렉션의 불변 조건도 완성시켰다는 생각이듬.
2. Pieces.java
- add() : 그에따른 접근 제한자 private 으로 변경

SANGCO reviewed (approved):
저는 미션4에서 더 이상 피드백 드릴게 없군요.
고생 많으셨습니다~

tmdgusya commented:
> 저는 미션4에서 더 이상 피드백 드릴게 없군요.
> 고생 많으셨습니다~

감사해요 좋은 피드백이였습니다.
덕분에 세심한 부분들에 대해서 공부할 수 있게 됬어요! 👍

tmdgusya committed:
fix : 3차 코드리뷰 반영

1. Pieces.java

- validateSize() : 적절한 표현 방식을 위한 변경

honux77 commented:
상로코치 모두 수고 많으셨어요! 리뷰 내용이 적절한 수준으로 잘 반영되고 있군요. Good! 🥇
